allennlp/tests/models/biaffine_dependency_parser_test.py

```diff
@@ -31,3 +33,30 @@ def test_decode_runs(self):
                                                       'tag_loss', 'loss', 'mask',
                                                       'predicted_dependencies', 'predicted_heads',
                                                       'words', 'pos'])
+
+    def test_mst_respects_no_outgoing_root_edges_constraint(self):
+        # This energy tensor expresses the following relation:
+        # energy[i,j] = "Score that j is the head of i". In this
+        # case, we have heads pointing to their children.
+
+        # We want to construct a case that has 2 children for the ROOT node,
+        # because in a typical dependency parse there should only be one
+        # word which has the ROOT as it's head.
+
+        energy = torch.Tensor([[0, 1, 1],
+                               [10, 0, 1],
+                               [10, 1, 0]]).view(1, 1, 3, 3)
+        # In this case, the maximum weight tree looks like this,
+        # with both edges having weight 10.
+        #               A(index 0)
+        #               /        \
+        #              \/        \/
+        #          B(index 1)   C(index 2)
+        length = torch.LongTensor([3])
+        heads, _ = self.model._run_mst_decoding(energy, length) # pylint: disable=protected-access
+        # This is the correct MST, but not desirable for dependency parsing.
+        assert heads.tolist()[0] == [-1, 0, 0]
+
+        energy[:, :, 0, :] = 0
```

Comment:
Is this example really doing what you think it's doing?

```python
>>> t = torch.Tensor([[0, 1, 1], [10, 0, 1], [10, 1, 0]]).view(1, 1, 3, 3)
>>> t
tensor([[[[  0.,   1.,   1.],
          [ 10.,   0.,   1.],
          [ 10.,   1.,   0.]]]])
>>> t[:, :, 0, :] = 0
>>> t
tensor([[[[  0.,   0.,   0.],
          [ 10.,   0.,   1.],
          [ 10.,   1.,   0.]]]])
```

This is zeroing out the top row, not the first column.  Is that what you expected?  And in general, I'd recommend using different numbers for every non-zero value, so there are no ties and it's more obvious what the MST should be.